src/utils/sshUtils.ts

```diff
@@ -3,29 +3,54 @@
 import { VirtualMachineTreeItem } from '../tree/VirtualMachineTreeItem';
+import { createComputeClient } from './azureClients';
 import { cpUtils } from "./cpUtils";
+import { nonNullValueAndProp } from './nonNull';
 
 export const sshFsPath: string = join(os.homedir(), '.ssh');
 
-export async function getSshKey(vmName: string, passphrase: string): Promise<string> {
+export async function getSshKey(context: IVirtualMachineWizardContext, vmName: string, passphrase: string): Promise<string> {
     return await callWithMaskHandling(
         async () => {
             const sshKeyName: string = `azure_${vmName}_rsa`;
             const sshKeyPath: string = join(sshFsPath, sshKeyName);
 
             if (!await fse.pathExists(`${sshKeyPath}.pub`)) {
-                ext.outputChannel.appendLog(localize('generatingKey', 'Generating public/private rsa key pair in "{0}"...', sshKeyPath));
+                const generatingKey: string = localize('generatingKey', 'Generating public/private rsa key pair in "{0}"...', sshKeyPath);
+                const generatedKey: string = localize('generatedKey', 'Generated public/private rsa key pair in "{0}".', sshKeyPath);
                 // create the .ssh folder if it doesn't exist
                 await fse.ensureDir(sshFsPath);
-                await cpUtils.executeCommand(undefined, undefined, 'ssh-keygen', '-t', 'rsa', '-b', '4096', '-f', cpUtils.wrapArgInQuotes(sshKeyPath), '-N', cpUtils.wrapArgInQuotes(passphrase));
-                ext.outputChannel.appendLog(localize('generatedKey', 'Generated public/private rsa key pair in "{0}".', sshKeyPath));
+                if (await sshKeygenExists()) {
+                    ext.outputChannel.appendLog(generatingKey);
+                    await cpUtils.executeCommand(undefined, undefined, 'ssh-keygen', '-t', 'rsa', '-b', '4096', '-f', cpUtils.wrapArgInQuotes(sshKeyPath), '-N', cpUtils.wrapArgInQuotes(passphrase));
+                    ext.outputChannel.appendLog(generatedKey);
+                } else {
+                    // if ssh-keygen isn't found, we can leverage Azure's SSH keys, but they do not allow for passphrases
+                    ext.outputChannel.appendLog(localize('creatingAzureKey', 'Unable to find `ssh-keygen` in environment variable PATH.  Creating new Azure SSH Key "{0}"...', vmName));
+                    if (passphrase) {
+                        ext.outputChannel.appendLog(localize('generatingKeyFailedNote', 'NOTE: Azure generated SSH keys do not support passphrases.'));
+                    }
+
+                    const client: ComputeManagementClient = await createComputeClient(context);
+                    const rgName: string = nonNullValueAndProp(context.resourceGroup, 'name');
+                    await client.sshPublicKeys.create(rgName, vmName, { location: nonNullValueAndProp(context.location, 'name') });
+                    ext.outputChannel.appendLog(localize('createdAzureKey', 'Created new Azure SSH Key "{0}".', vmName));
```

Comment:
I'm confused on the difference between `sshPublicKeys.create` and `sshPublicKeys.generateKeyPair` and I feel like these logs might be confusing to the user, too. What exactly is going on between those two? If the "create" thing is creating some resource in Azure, should we be deleting it afterwards?

Reply:
Yeah, create is creating the Azure resource, but it doesn't actually contain a public key yet.  Generate key is a one time action that creates a private/public key pair and returns that response.

After it's generated, the public key can be retrieved from the Azure resource if the user ever wanted it again.  This information would also be available in the VM itself so I'm not against deleting it afterwards

![image](https://user-images.githubusercontent.com/5290572/102653968-9dfa8080-4124-11eb-9a23-13418869d4d9.png)

Reply:
I mean you didn't even mention users can get the public key from the ".pub" file on their local machine, which matches ssh-keygen and is the most important place IMO. I think I would prefer if we deleted the resource afterwards. I feel like users who fall through to this case will only be confused by a magic ssh resource in Azure.

Reply:
In that case, it might be worth just hiding the Azure SSH key and just create it that way.  But I still need to convey to them that their passphrase won't be expected (though I doubt that'll happen too often)

Reply:
I don't know what you mean by "hiding the Azure SSH key"

Reply:
Hiding the creation of it. Keep the message the same that we're `Generating public/private rsa key pair`, but alert the user that the passphrase isn't respected because we couldn't detect ssh-keygen?

Reply:
Oh - no strong preference on the notifications/logs